lib/Bridge/TolerantParser/WorseReflection/DoctrineAnnotationCompletor.php

```diff
@@ -0,0 +1,139 @@
+use Phpactor\WorseReflection\Reflector;
+
+class DoctrineAnnotationCompletor extends NameSearcherCompletor implements Completor
+{
+    /**
+     * @var Reflector
+     */
+    private $reflector;
+
+    /**
+     * @var Parser
+     */
+    private $parser;
+
+    public function __construct(
+        NameSearcher $nameSearcher,
+        Reflector $reflector,
+        Parser $parser = null
+    ) {
+        parent::__construct($nameSearcher);
+
+        $this->reflector = $reflector;
+        $this->parser = $parser ?: new Parser();
+    }
+
+    public function complete(TextDocument $source, ByteOffset $byteOffset): Generator
+    {
+        $truncatedSource = $this->truncateSource((string) $source, $byteOffset->toInt());
+        $sourceNodeFile = $this->parser->parseSourceFile((string) $source);
+
+        $node = $this->findNodeForPhpdocAtPosition(
+            $sourceNodeFile,
+            // the parser requires the byte offset, not the char offset
+            strlen($truncatedSource)
+        );
+
+        if (!$node) {
+            // Ignore this case is the cursor is not in a phpdoc block
+            return true;
+        }
+
+        if (!$annotation = $this->extractAnnotation($truncatedSource)) {
+            // Ignore if not an annotation
+            return true;
+        }
+
+        $suggestions = $this->completeName($annotation);
+
+        foreach ($suggestions as $suggestion) {
+            if (!$this->isAnAnnotation($suggestion)) {
+                continue;
+            }
+
+            yield $suggestion;
+        }
+
+        return $suggestions->getReturn();
+    }
+
+    protected function createSuggestionOptions(NameSearchResult $result): array
+    {
+        return array_merge(parent::createSuggestionOptions($result), [
+            'snippet' => (string) $result->name()->head() .'($1)$0',
+        ]);
+    }
+
+    private function truncateSource(string $source, int $byteOffset): string
+    {
+        // truncate source at byte offset - we don't want the rest of the source
+        // file contaminating the completion (for example `$foo($<>\n    $bar =
+        // ` will evaluate the Variable node as an expression node with a
+        // double variable `$\n    $bar = `
+        $truncatedSource = substr($source, 0, $byteOffset);
+
+        // determine the last non-whitespace _character_ offset
+        $characterOffset = OffsetHelper::lastNonWhitespaceCharacterOffset($truncatedSource);
+
+        // truncate the source at the character offset
+        $truncatedSource = mb_substr($source, 0, $characterOffset);
+
+        return $truncatedSource;
+    }
+
+    private function findNodeForPhpdocAtPosition(SourceFileNode $sourceNodeFile, int $position): ?Node
+    {
+        /** @var Node $node */
+        foreach ($sourceNodeFile->getDescendantNodes() as $node) {
+            if (
+                $node->getFullStart() < $position
+                && $position < $node->getStart()
+            ) {
```

Comment:
Not important: I would make this a guard condition:

```
if ($node->getFullStart() >= $position ||  $position >= $node->getStart()) {
    continue;
}

return $node->getDocCommentText() ? $node : null;
```

Reply:
In most cases I would agree.
Here I think the cognitive load is a lot more important,the condition being in a loop does not help, in order to know when you return you also have to negate the condition from the if.
Which mean you have to inverse all the logic instead of reading it.
> If the position is between the fullstart and start position of a node then it's the node I want

Looks easier to grasp than

> If the position is not before full start nor after start then it's the node I want

Its up to you though but then I would put the result of the condition into a variable to make it clear.

Reply:
I think it decreases the load, as you can clearly see what is out-of-bounds, and anyway - I still don't understand exactly what it's :smile: 

Fine to leave it as is.